src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/LSPRazorProjectHost.cs

```diff
@@ -0,0 +1,115 @@
+    internal class LSPRazorProjectHost : IProjectDynamicLoadComponent
+    {
+        private readonly LSPEditorFeatureDetector _featureDetector;
+        private readonly List<Lazy<ILanguageClient, ILanguageClientMetadata>> _applicableLanguageClients;
+        private readonly Lazy<ILanguageClientBroker> _languageClientBroker;
+
+        [ImportingConstructor]
+        public LSPRazorProjectHost(
+            LSPEditorFeatureDetector featureDetector,
+            Lazy<ILanguageClientBroker> languageClientBroker,
+            [ImportMany] IEnumerable<Lazy<ILanguageClient, IDictionary<string, object>>> languageClients)
+        {
+            if (featureDetector is null)
+            {
+                throw new ArgumentNullException(nameof(featureDetector));
+            }
+
+            if (languageClientBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageClientBroker));
+            }
+
+            if (languageClients is null)
+            {
+                throw new ArgumentNullException(nameof(languageClients));
+            }
+
+            _featureDetector = featureDetector;
+            _languageClientBroker = languageClientBroker;
+            _applicableLanguageClients = GetApplicableClients(languageClients);
+        }
+
+        public Task LoadAsync()
+        {
+            if (!_featureDetector.IsLSPEditorFeatureEnabled())
+            {
+                return Task.CompletedTask;
+            }
+
+            foreach (var client in _applicableLanguageClients)
+            {
+                _languageClientBroker.Value.LoadAsync(client.Metadata, client.Value).Forget();
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task UnloadAsync()
+        {
+            return Task.CompletedTask;
+        }
+
+        private static List<Lazy<ILanguageClient, ILanguageClientMetadata>> GetApplicableClients(IEnumerable<Lazy<ILanguageClient, IDictionary<string, object>>> languageClients)
+        {
+            var applicableContentTypes = new[]
+            {
+                RazorLSPContentTypeDefinition.Name,
+                CSharpVirtualDocumentFactory.CSharpLSPContentTypeName,
+                HtmlVirtualDocumentFactory.HtmlLSPContentTypeName,
+            };
+
+            var applicableClients = new List<Lazy<ILanguageClient, ILanguageClientMetadata>>();
+            foreach (var client in languageClients)
+            {
+                if (!client.Metadata.TryGetValue(nameof(ILanguageClientMetadata.ContentTypes), out object contentTypeValue) ||
+                    !(contentTypeValue is IEnumerable<string> contentTypes) ||
+                    !contentTypes.Intersect(applicableContentTypes).Any())
+                {
+                    continue;
+                }
+
+                string clientName = null;
+                if (client.Metadata.TryGetValue(nameof(ILanguageClientMetadata.ClientName), out object clientNameValue))
```

Comment:
Hmmm, this definitely makes me squirm a bit. Any additional. feature that the LoadAsync method enables will fall over here. Is there a better way we can load without extracting all metadata the LoadAsync could care about?

Reply:
We could with reflection. But without it, I believe this is the only way. There is also another problem with this which I'm not sure we can solve. [LanguageClientMetadata](https://devdiv.visualstudio.com/DevDiv/_git/VSLanguageServerClient?path=%2Fsrc%2Fproduct%2FRemoteLanguage%2FImpl%2FLanguageClientMetadata.cs&version=GBdevelop&line=14&lineEnd=15&lineStartColumn=1&lineEndColumn=1&lineStyle=plain) has more properties than what is publicly exposed. So if the servers set more than `ClientName` and ` ContentTypes`, that information would be lost when we convert the `IDictionary` to `ILanguageClientMetadata`.